minio/credentials/providers.py

```diff
@@ -482,9 +482,8 @@ def retrieve(self):
         res = _urlopen(
             self._http_client,
             "POST",
-            self._sts_endpoint,
-            body=self._body,
-            headers={"Content-Type", "application/x-www-form-urlencoded"},
+            self._sts_endpoint + "?" + self._url, # looks like encoded url should be used to form full url, instead of passed as a body ...
+            headers={"Content-Type": "application/x-www-form-urlencoded"}, # here should be dict instead of set
```

Comment:
This is actually a fix. It is a typo `,` instead of `:`. Rest of the changes in this PR are violation of specification.

Reply:
_Ok. I don't know what your specifications are. But, what I could assure you, is that the latest version of minio server installed on my machine returns code 400 without:_

self._sts_endpoint + "?" + self._url

_Second, you could also try to debug you minio-go client, especially Retrieve method of LDAPIdentity struct. Look what is returned by the u.String(), after you assign:_

u.RawQuery = v.Encode()

_at line 86 of sts_ldap_identity.go file_

Reply:
This is a server behavior problem, LDAP based STS expects as URL instead of body unlike other STS implementations.

Reply:
Its because the server wrongly restricts based on query params @stepanko7

Reply:
A minor fix on server side will fix it, until then lets not make such changes in minio-py

```diff
 git diff
diff --git a/cmd/sts-handlers.go b/cmd/sts-handlers.go
index 152536e29..80f86b30e 100644
--- a/cmd/sts-handlers.go
+++ b/cmd/sts-handlers.go
@@ -274,6 +274,9 @@ func (sts *stsAPIHandlers) AssumeRoleWithJWT(w http.ResponseWriter, r *http.Requ
 
        action := r.Form.Get(stsAction)
        switch action {
+       case ldapIdentity:
+               sts.AssumeRoleWithLDAPIdentity(w, r)
+               return
        case clientGrants, webIdentity:
        default:
                writeSTSErrorResponse(ctx, w, true, ErrSTSInvalidParameterValue, fmt.Errorf("Unsupported action %s", action))
```

Reply:
> _Ok. I don't know what your specifications are. But, what I could assure you, is that the latest version of minio server installed on my machine returns code 400 without:_
> 
> self._sts_endpoint + "?" + self._url
> 
> _Second, you could also try to debug you minio-go client, especially Retrieve method of LDAPIdentity struct. Look what is you request (req) at line 88, after you assign:_
> 
> u.RawQuery = v.Encode()
> 
> _at line 86 of sts_ldap_identity.go file_

Understood. Now you would need to fix differently.

```diff
diff --git a/minio/credentials/providers.py b/minio/credentials/providers.py
index 195635a..28b17e3 100644
--- a/minio/credentials/providers.py
+++ b/minio/credentials/providers.py
@@ -464,15 +464,7 @@ class LdapIdentityProvider(Provider):
     def __init__(
             self, sts_endpoint, ldap_username, ldap_password, http_client=None,
     ):
-        self._sts_endpoint = sts_endpoint
-        self._http_client = http_client or urllib3.PoolManager(
-            retries=urllib3.Retry(
-                total=5,
-                backoff_factor=0.2,
-                status_forcelist=[500, 502, 503, 504],
-            ),
-        )
-        self._body = urlencode(
+        self._sts_endpoint = sts_endpoint + "?" + urlencode(
             {
                 "Action": "AssumeRoleWithLDAPIdentity",
                 "Version": "2011-06-15",
@@ -480,6 +472,13 @@ class LdapIdentityProvider(Provider):
                 "LDAPPassword": ldap_password,
             },
         )
+        self._http_client = http_client or urllib3.PoolManager(
+            retries=urllib3.Retry(
+                total=5,
+                backoff_factor=0.2,
+                status_forcelist=[500, 502, 503, 504],
+            ),
+        )
         self._credentials = None
 
     def retrieve(self):
@@ -492,8 +491,6 @@ class LdapIdentityProvider(Provider):
             self._http_client,
             "POST",
             self._sts_endpoint,
-            body=self._body,
-            headers={"Content-Type", "application/x-www-form-urlencoded"},
         )
 
         self._credentials = _parse_credentials(
```

Reply:
@harshavardhana 
> A minor fix on server side will fix it, until then lets not make such changes in minio-py
> 

Current behaviour of Minio server behaves exactly same as `AssumeRoleWithClientGrants` and `AssumeRoleWithWebIdentity`; and changing server would break backward compatibility as well. I think fixing the client is right thing.

Reply:
It won't break backward compatibility, server supports both query based as well as form Post this is a requirement as per STS API requirements.

ClientGrants and WebIdentity support both, we just have to support LDAP to also support form body.

Reply:
@harshavardhana :+1: I think we can go ahead and take this PR with the change I suggested

Reply:
> @harshavardhana :+1: I think we can go ahead and take this PR with the change I suggested

Okay